src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedStaticConstructor.cs

```diff
@@ -371,5 +372,69 @@ internal override int CalculateLocalSyntaxOffset(int localPosition, SyntaxTree l
             var containingType = (SourceMemberContainerTypeSymbol)this.ContainingType;
             return containingType.CalculateSyntaxOffsetInSynthesizedConstructor(localPosition, localTree, isStatic: true);
         }
+
+#nullable enable
+        private readonly object _syncRoot = new object();
+        private Binder.ProcessedFieldInitializers _processedInitializers;
```

Comment:
>Binder.ProcessedFieldInitializers _processedInitializers; [](start = 16, length = 57)

It is a bad idea to cache this on the symbol. Perhaps, I wasn't clear during our offline conversation. I  was talking about caching a bool representing whether to emit the method or not. I don't think we should be concerned much about reusing result of binding. The reason why caching Binder.ProcessedFieldInitializers is a bad idea is that it holds on to bound nodes, those refer to syntax nodes, making it impossible to unload syntax trees. IDE will be very unhappy. Symbols should never keep syntax nodes alive, unless the symbols are very short lived.

Reply:
Was the thought for it to work something like this?

```cs
private bool? _shouldEmit;
internal bool ShouldEmit(ImmutableArray<BoundInitializer> boundInitializersOpt = default)
{
    if (_shouldEmit != null)
        return _shouldEmit.GetValueOrDefault();

    if (boundInitializersOpt.IsDefault)
        boundInitializersOpt = BindFieldInitializers(...);

    _shouldEmit = CheckFieldInitializers(boundInitializersOpt, ...);
    return _shouldEmit.GetValueOrDefault();
}
```

Reply:
> Was the thought for it to work something like this?

Yes, but bool? is not thread-safe, we usually using TreeState enim or something like that, we have a type and examples


---
In reply to: [413184581](https://github.com/dotnet/roslyn/pull/43234#discussion_r413184581) [](ancestors = 413184581)